extras/rebalance/directory-rebalance.py

```diff
@@ -0,0 +1,203 @@
+                        follow_symlinks=False)
+            return True, None #Indicate that migration happened
+
+        except OSError as e:
+            return False, e
+
+    #Updates the total,migrated,skipped files/size and durations of the migrations
+    def migrate_with_stats(self, f, size):
+        migration_start = time.perf_counter()
+        result, err = self.migrate_data(f)
+        migration_end = time.perf_counter()
+        duration = migration_end - migration_start
+
+        if err is not None:
+            if err.errno == errno.EEXIST:
+                logging.info("{} - Not needed".format(f))
+            elif err.errno == errno.ENOENT:
+                logging.info("{} - file not present anymore".format(f))
+            else:
+                logging.critical("{} - {} - exiting.".format(f, err))
+                self.exit()
+                sys.exit(1)
+
+        if result and duration != 0:
+            logging.info("{} - {} [{}] - {}/s".format(f, size_fmt(size), size, size_fmt(size/duration)))
+
+        self.total_files += 1
+        self.total_size += size
+        if result == True:
+            self.migrated_files += 1
+            self.migrated_size += size
+            self.migration_duration += duration
+        else:
+            self.skipped_migration_duration += duration
+
+        return result
+
+    def run(self):
+        print("Starting Rebalance")
+        self.rebalance_start = time.perf_counter()
+        with open(self.index) as f:
+            sample_size = 1024*1024
+            sample_files = 100
+            i = 1
+            for line in f:
+                file_path, _, file_size = line.rpartition('-')
+                file_size = int(file_size)
+                self.migrate_with_stats(file_path, file_size)
+                if self.migrated_size > i*sample_size and self.migrated_files > i*sample_files:
+                    i += 1
+                    now = time.perf_counter()
+                    duration = now - self.rebalance_start
+                    if duration != 0.0:
+                        speed = self.migrated_size/duration
+                        migration_fraction = self.migrated_size/self.total_size
+                        eta = ((self.expected_total_size - self.total_size)/speed)*migration_fraction
+                        progress(self.total_size, self.expected_total_size, "ETA: {}".format(time_fmt(eta)))
+
+        self.exit()
+
+    #For each file in the directory recursively, writes <file-path>-<size> to index file
+    def generate_rebalance_file_index(self):
+        with open(self.index, 'w') as file_index:
+            total_size = 0
+            for d, _, files in os.walk(self.path):
+                for f in files:
+                    try:
+                        path = os.path.join(d, f)
+                        size = os.stat(path).st_size
+                        file_index.write(path+'-'+str(size)+'\n')
```

Comment:
Have you considered that the file size may change between the time the list is generated and the time the rebalance is actually done ? this could have a significant impact in statistics in some cases.

Reply:
I missed that :-(. What do you suggest I do in this case? I would prefer not to do stat call again.

Reply:
Except this one, I addressed the rest. The only option that I find acceptable is: Do one more stat and if the size changes, then adjust expected_total_size to reflect that and take it from there. If there is an ENOENT we should decrement that size too. Let me know what you think about this one.

Reply:
Yes, I think this is the most accurate approach if the volume is under use.